sklearn/cluster/_kmeans.py

```diff
@@ -1635,6 +1645,80 @@ def _check_params(self, X):
                 f"reassignment_ratio should be >= 0, got "
                 f"{self.reassignment_ratio} instead.")
 
+    def _mini_batch_convergence(self, step, n_steps, n_samples,
+                                centers_squared_diff, batch_inertia):
+        """Helper function to encapsulate the early stopping logic"""
+        # Normalize inertia to be able to compare values when
+        # batch_size changes
+        batch_inertia /= self._batch_size
+
+        # count steps starting from 1 for user friendly verbose mode.
+        step = step + 1
+
+        # Ignore first iteration because it's inertia from initialization.
+        if step == 1:
+            if self.verbose:
+                print(f"Minibatch step {step}/{n_steps}: mean batch "
+                      f"inertia: {batch_inertia}")
+            return False
+
+        # Compute an Exponentially Weighted Average of the inertia to
+        # monitor the convergence while discarding minibatch-local stochastic
+        # variability: https://en.wikipedia.org/wiki/Moving_average
+        if self._ewa_inertia is None:
+            self._ewa_inertia = batch_inertia
+        else:
+            alpha = self._batch_size * 2.0 / (n_samples + 1)
+            alpha = min(alpha, 1)
+            self._ewa_inertia = (
+                self._ewa_inertia * (1 - alpha) + batch_inertia * alpha)
+
+        # Log progress to be able to monitor convergence
+        if self.verbose:
+            print(f"Minibatch step {step}/{n_steps}: mean batch inertia: "
+                  f"{batch_inertia}, ewa inertia: {self._ewa_inertia}")
+
+        # Early stopping based on absolute tolerance on squared change of
+        # centers position
+        if self._tol > 0.0 and centers_squared_diff <= self._tol:
+            if self.verbose:
+                print(f"Converged (small centers change) at step "
+                      f"{step}/{n_steps}")
+            return True
```

Comment:
It would be great to find a way to trigger this case. It is currently not covered by the tests.

Reply:
I added a test ``test_minibatch_declared_convergence``

Reply:
Good test!